c/dao.c

```diff
@@ -130,104 +145,170 @@ static int extract_withdraw_header_index(size_t input_index, size_t *index) {
-  mol_pos_t deposit_dao_pos;
-  len = HEADER_SIZE;
-  ret = ckb_load_header(deposit_header_buffer, &len, 0, input_index,
-                        CKB_SOURCE_INPUT);
-  if (ret != CKB_SUCCESS) {
-    return ret;
-  }
-  ret = extract_header_number_and_dao(deposit_header_buffer, len, &deposit_number, &deposit_dao_pos);
+  dao_header_data_t withdraw_data;
+  ret = load_dao_header_data(withdraw_index, CKB_SOURCE_HEADER_DEP,
+                             &withdraw_data);
   if (ret != CKB_SUCCESS) {
     return ret;
   }
 
-  unsigned char withdraw_header_buffer[HEADER_SIZE];
-  uint64_t withdraw_number;
-  mol_pos_t withdraw_dao_pos;
-  len = HEADER_SIZE;
-  ret = ckb_load_header(withdraw_header_buffer, &len, 0, withdraw_index,
-                        CKB_SOURCE_HEADER_DEP);
-  if (ret != CKB_SUCCESS) {
-    return ret;
-  }
-  ret = extract_header_number_and_dao(withdraw_header_buffer, len, &withdraw_number, &withdraw_dao_pos);
+  dao_header_data_t deposit_data;
+  ret = load_dao_header_data(input_index, CKB_SOURCE_INPUT, &deposit_data);
   if (ret != CKB_SUCCESS) {
     return ret;
   }
 
-  if (withdraw_number <= deposit_number) {
+  uint64_t withdraw_fraction =
+      withdraw_data.epoch_index * deposit_data.epoch_length;
+  uint64_t deposit_fraction =
+      deposit_data.epoch_index * withdraw_data.epoch_length;
+  if ((withdraw_data.epoch_number < deposit_data.epoch_number) ||
+      ((withdraw_data.epoch_number == deposit_data.epoch_number) &&
+       (withdraw_fraction <= deposit_fraction))) {
     return ERROR_INVALID_WITHDRAW_BLOCK;
   }
 
-  uint64_t windowleft = LOCK_PERIOD_BLOCKS -
-                        (withdraw_number - deposit_number) % LOCK_PERIOD_BLOCKS;
-  uint64_t minimal_since =
-      withdraw_number + MAX(MATURITY_BLOCKS, windowleft) + 1;
+  uint64_t deposited_epoches =
+      withdraw_data.epoch_number - deposit_data.epoch_number;
+  /*
+   * This is essentially a round-up operation. Suppose withdraw epoch is
+   * a + b / c, deposit epoch is d + e / f, the deposited epoches will be:
+   *
+   * (a - d) + (b / c - e / f) == (a - d) + (b * f - e * c) / (c * f)
+   *
+   * If (b * f - e * c) is larger than 0, we will have a fraction part in
+   * the deposited epoches, we just add one full epoch to deposited_epoches
+   * to round it up.
+   * If (b * f - e * c) is no larger than 0, let's look back at (b / c - e / f),
+   * by the definition of a fraction, we will know 0 <= b / c < 1, and
+   * 0 <= e / f < 1, so we will have -1 < (b / c - e / f) <= 0, hence
+   * (a - d) - 1 < (a - d) + (b / c - e / f) <= (a - d), we won't need to do
+   * anything for a round-up operation.
+   */
+  if (withdraw_fraction > deposit_fraction) {
+    deposited_epoches++;
+  }
+  uint64_t lock_epoches = (deposited_epoches + (LOCK_PERIOD_EPOCHES - 1)) /
+                          LOCK_PERIOD_EPOCHES * LOCK_PERIOD_EPOCHES;
+  /* Cell must at least be locked for one full lock period(180 epoches) */
+  if (lock_epoches < LOCK_PERIOD_EPOCHES) {
```

Comment:
Indeed, this condition never holds

Reply:
I thought about this as well, and I just leave it there as a final safeguard in case there's something that we miss. From a cycle point of view, this only cost us several cycles which can be ignored.